### PR TITLE
Preferences: Memoize pref handlers, add styling

### PR DESCRIPTION
**Pull request by mattgawarecki**
#### Changes proposed in this Pull Request

* Encapsulate the logic to choose a correct preference handler in a `useMemo`, to reduce unnecessary re-renders when the Preferences Helper is visible.
* Encapsulate the `unsetPreference` action in a `useCallback`, to reduce unnecessary re-renders when the Preferences Helper is visible.
* Make a few `Preference` components functional (as opposed to class-based).
* Re-style and add translations to the Preferences Helper to make it a little friendlier. 🙂 

#### Testing instructions

* In a local testing environment, start Calypso Blue.
* Open the Preferences Helper by going to the bottom right corner, hovering over the environment name, and then hovering over **PREFERENCES** in the menu that pops out.
* Verify that you're able to manage your user preferences just as in production: deleting should be possible for all items, and altering should be possible for all but the most complex data types. Be sure to test and compare the visibility and behavior of the "save" and "reset" buttons for preference types where they're applicable.

##### Before

Screen Shot 2021-06-21 at 14 53 45

##### After

Screen Shot 2021-06-21 at 14 53 02

### CODE REVIEW

File: client/lib/preferences-helper/string-preference.tsx [hunk 1]
```diff
@@ -45,18 +47,18 @@ const StringPreference: FunctionComponent< Props > = ( { name, value } ) => {
...
+						{ translate( 'save' ) }
+					</button>
 					<button
-						className="preferences-helper__reset-pref-button"
+						className="preferences-helper__preference-reset"
 						onClick={ resetPreferenceChange }
 						disabled={ value === localValue }
 					>
-						{ 'reset' }
+						{ translate( 'reset' ) }
```
Comment (a8ci18n) [inline client/lib/preferences-helper/string-preference.tsx:61]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **10** times:
`translate( 'Reset ' )` *ES Score: 9*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/fix/unnecessary-renders-in-prefs-helper)

**Review by cbauerman — CHANGES_REQUESTED** [review]
Screen Shot 2021-06-21 at 2 20 38 PM

Something changed styling-wise, this window shouldn't be _quite_ so wide 😛

**Review by tyxla — COMMENTED** [review]
I haven't reviewed it thoroughly yet, but several questions and thoughts come to mind at first glance:

* I really appreciate the Lodash refactors 🙌 
* Do we really want to add translations here? The preferences helper (and all of the environment badge tools) are not available in production, so wouldn't that be a waste of time and money for translation? 
* If we're overhauling the visual part, is it a good time to adopt `@wordpress/components`? Would that save us most of the styles necessary?
* Going through the code and the PR description, it seems like this one includes several PRs. Does it make sense to split it into better-defined logical pieces of changes?

File: client/lib/preferences-helper/boolean-preference.tsx [hunk 2]
```diff
@@ -45,18 +47,18 @@ const BooleanPreference: FunctionComponent< Props > = ( { name, value } ) => {
 			{ value !== localValue && (
 				<>
 					<button
-						className="preferences-helper__save-pref-button"
+						className="preferences-helper__preference-save"
 						onClick={ savePreferenceChange }
 						disabled={ value === localValue }
```
Comment (jsnajdr) [inline client/lib/preferences-helper/boolean-preference.tsx:52]: This `disabled` prop is always `false` and can be omitted, because the entire buttons fragment is rendered only when `value !== localValue`

File: client/lib/preferences-helper/boolean-preference.tsx [hunk 3]
```diff
@@ -45,18 +47,18 @@ const BooleanPreference: FunctionComponent< Props > = ( { name, value } ) => {
 			{ value !== localValue && (
 				<>
 					<button
-						className="preferences-helper__save-pref-button"
+						className="preferences-helper__preference-save"
```
Comment (jsnajdr) [inline client/lib/preferences-helper/boolean-preference.tsx:50]: Does the `__preference-save` need to always contain the `preference` word? It's already in the prefix, so the class names seem too wordy to me.

File: client/lib/preferences-helper/preference-list.js [hunk 4]
```diff
@@ -14,41 +13,55 @@ import { getAllRemotePreferences } from 'calypso/state/preferences/selectors';
...
+	const translate = useTranslate();
 
-export default connect(
-	( state ) => ( {
-		preferences: getAllRemotePreferences( state ),
-	} ),
-	null
-)( localize( PreferenceList ) );
+	const preferences = useSelector( getAllRemotePreferences );
+	const sortedPreferenceItems = useMemo( () => {
```
Comment (jsnajdr) [inline client/lib/preferences-helper/preference-list.js:20]: Does the `useMemo` perform any useful optimization here? I think it doesn't. The `PreferenceList` component doesn't have any props and rerenders only when the `preferences` selector returns a new value. But then `sortedPreferenceItems` needs to be re-calculated, despite the `useMemo`. If the `preferences` value doesn't change, `PreferenceList` is never rerendered and the memoized value never has a chance to be used.

**Comment by jsnajdr** [comment]
> If we're overhauling the visual part, is it a good time to adopt `@wordpress/components`?

This is something I was experimenting with when writing the `AuthHelper` for switching between OAuth and cookies. The Calypso helpers' UI is very primitive and would deserve to be rewritten using standard components: popovers, forms...

**Comment by sarayourfriend** [comment]
If you end up deciding to re-write using `@wordpress/components`, feel free to reach out to the @Automattic/team-calypso-components squad ❤️

**Comment by mattgawarecki** [comment]
> * Going through the code and the PR description, it seems like this one includes several PRs. Does it make sense to split it into better-defined logical pieces of changes?

On reflection, I think @tyxla's right. I'll be taking everyone's feedback into account here, but I'm closing this PR and making some more focused ones to replace it.
